design/src/rfcs/rfc0012_bytestream_callback_api.md

```diff
@@ -0,0 +1,112 @@
+RFC: A callback API for `ByteStream`
+====================================
+
+> Status: RFC
+
+Adding a callback API to `ByteStream` will enable developers using the SDK to implement things like checksum validations and 'read progress' callbacks.
+
+## The Implementation
+
+*Note that comments starting with '//' are not necessarily going to be included in the actual implementation and are intended as clarifying comments for the purposes of this RFC.*
+
+```rust
+// Each trait method defaults to doing nothing. It's up to implementors to
+// implement one or both of the trait methods
+/// Structs and enums implementing this trait can be inserted into a `ByteStream`,
+/// and will then be called in reaction to various events during a `ByteStream`'s
+/// lifecycle.
+pub trait ByteStreamReadCallback
+// I don't really like this bound but we need it because this will be inserted into an `Inner`
+where
+    Self: Debug + Clone + PartialEq + Eq,
+{
+    /// This callback is called for each chunk **successfully** read.
+    /// If an error occurs while reading a chunk, this will not be called.
+    /// This function takes `&mut self` so that implementors may modify
+    /// an implementing struct/enum's internal state.
+    // In order to stop the compiler complaining about these empty default impls,
+    // we allow unused variables.
+    fn on_read_chunk(&mut self, #[allow(unused_variables)] chunk: &Bytes) {}
+
+    /// This callback is called once all chunks have been successfully read.
+    /// It's passed a reference to the chunks in the form of an `AggregatedBytes`.
+    /// This function takes `&mut self` so that implementors may modify an
+    /// implementing struct/enum's internal state.
+    fn finally(&mut self, #[allow(unused_variables)] aggregated_bytes: &AggregatedBytes) {}
+}
+
+// We add a new method to `ByteStream` for inserting callbacks
+impl ByteStream {
+    // ...other impls omitted
+
+    // Read callbacks can only be inserted, not removed or reordered. If users
+    // desire extra management functions, we can add them in a later update.
+    // Callbacks are actually stored and called from the `Inner` object
+    pub fn insert_read_callback(&mut self, callback: Box<dyn ByteStreamReadCallback>) {
```

Comment:
If we don't have remove, then I think the verb shouldn't be insert. The `with_callback` function is probably sufficient.